Patch 1

FlyinPhil opened the pull request:
Ajout entre parenthèses de la vitesse réelle avec la tolérance de recouvrement des voiles (si on est en voile automatique, le changement de voile ne se fait pas si la vitesse de la meilleure voile est inférieur à la vitesse de la voile actuelle * 1.014. De plus, le bateau profite de la vitesse de la meilleure voile.)
J'ai dû élargie un peu les colonnes de chaque coté du graph pour faire tenir cette info.
et dans polars_app.js, ajout d'une constante "sailTolérance", le plus propre serait de prendre la valeur "autoSailChangeTolerance" dans le fichier json des polaires de chaque bateau. Mais j'ai fait au plus simple, car elle est identique pour tous les bateaux et n'a pas changé depuis 2 ans.

FlyinPhil committed:
Update general.css

FlyinPhil committed:
Update index.htm

FlyinPhil committed:
Update polars_app.js

Ajout entre parenthèses de la vitesse réelle avec la tolérance de recouvrement des voiles (si on est en voile automatique, le changement de voile ne se fait pas si la vitesse de la meilleure voile est inférieur à la vitesse de la voile actuelle * 1.014. De plus, le bateau profite de la vitesse de la meilleure voile.)

toxcct commented on css/general.css:
```diff
@@ -54,7 +54,7 @@ footer a {color:#8c8cb8; text-decoration:underline;}
 
 /**** CONTENT *******/
-article							{width:800px; cursor:default;}
+article							{width:850px; cursor:default;}
```
Pas favorable à élargir la largeur par défaut, à cause du support de la largeur la plus répandue sur les supports mobiles.

toxcct commented on css/general.css:
```diff
@@ -65,7 +65,7 @@ article .sidebox.grid1cols				{display:grid; grid-gap:2px; grid-template-columns
 article .sidebox.grid2cols.c20x1fr		{display:grid; grid-gap:2px; grid-template-columns:20px 1fr;			}
 article .sidebox.grid3cols				{display:grid; grid-gap:2px; grid-template-columns:20px 1fr 20px;		}
-article .sidebox.grid3cols.c20x1frx35	{display:grid; grid-gap:2px; grid-template-columns:20px 1fr 35px;		}
+article .sidebox.grid3cols.c20x1frx85	{display:grid; grid-gap:2px; grid-template-columns:20px 1fr 85px;		}
```
Pas favorable non plus à remplacer une classe CSS par une autre dans sa définition.
Je préfèrerais que la nouvelle classe soit ajoutée en parallèle de la précédente.

toxcct commented on css/general.css:
```diff
@@ -75,7 +75,7 @@ article .sidebox.grid5cols	> .title	{grid-column:span 5;}
 article .sidebox.grid2cols	> .sep		{grid-column:span 2;}
 article .sidebox.grid5cols	> .sep		{grid-column:span 5;}
-article .sidebox.grid5cols	> input		{grid-column:span 3; text-align:center; vertical-align:middle; height:18px; border:1px solid #999; padding:0;}
+article .sidebox.grid5cols	> input		{grid-column:span 2; text-align:center; vertical-align:middle; height:18px; border:1px solid #999; padding:0;}
```
Pourquoi ?

FlyinPhil replied:
Problème d'affichage sur Firefox après largeur passée à 850

toxcct replied:
Etonnant. Je vérifierai ce point.
En attendant, on ne va pas l'appliquer car c'est une rustine assez moche et il y a sûrement une meilleure façon de faire avec les largeurs de la grid...

FlyinPhil replied:
ok

toxcct replied:
J'ai trouvé un fix pour ce bug dans Firefox (que je n'explique pas vraiment).
C'est de fixer la largeur du champ à 100% de la zone qu'il est censé remplir (moins la largeur du gap entre 2 colonnes de la grid) :

Dans la classe `article .sidebox.grid5cols > input` (general.css:78), il faut y ajouter `width: calc(100% - 2px);`

FlyinPhil replied:
ok

toxcct commented on index.htm:
```diff
@@ -141,6 +141,7 @@
 				<label class="input_lbl" title="True Wind Speed" for ="data_tws">TWS:</label>
 				<input type="text" id="data_tws" maxlength="5" class="val unit_kts center" value="0" autofocus />
 				<span class="input_lbl unit_kts"></span>
+				<span>&nbsp;</span>
```
Je comprends que cet ajout est lié à la modif de la ligne 78 du fichier general.css, mais je ne comprends quand même pas le but de ce rétrécissement...

toxcct replied:
```suggestion
```

toxcct commented on css/general.css:
```diff
@@ -65,7 +65,7 @@ article .sidebox.grid1cols				{display:grid; grid-gap:2px; grid-template-columns
 article .sidebox.grid2cols.c20x1fr		{display:grid; grid-gap:2px; grid-template-columns:20px 1fr;			}
 article .sidebox.grid3cols				{display:grid; grid-gap:2px; grid-template-columns:20px 1fr 20px;		}
-article .sidebox.grid3cols.c20x1frx35	{display:grid; grid-gap:2px; grid-template-columns:20px 1fr 35px;		}
+article .sidebox.grid3cols.c20x1frx85	{display:grid; grid-gap:2px; grid-template-columns:20px 1fr 85px;		}
```
Pas favorable non plus à remplacer une classe CSS par une autre dans sa définition.
Je préfèrerais que la nouvelle classe soit ajoutée en parallèle de la précédente.

FlyinPhil replied:
Ok, j'ai vérifié qu'elle n'était pas utilisée ailleurs mais je comprends, je vais remettre la c20x1frx35.

toxcct replied:
Oui, l'idée est d'anticiper l'avenir, et de basculer d'une classe à l'autre en fonction de la largeur de la fenêtre disponible.

A voir comment cela se comporte également avec un `grid-template-columns:20px 1fr 1fr;` au lieu de fixer la largeur explicitement à `85px`...

toxcct replied:
```suggestion
article .sidebox.grid3cols.c20x1frx35	{display:grid; grid-gap:2px; grid-template-columns:20px 1fr 35px;		}
article .sidebox.grid3cols.c20x1frx85	{display:grid; grid-gap:2px; grid-template-columns:20px 1fr 85px;		}
```

toxcct commented on css/general.css:
```diff
@@ -54,7 +54,7 @@ footer a {color:#8c8cb8; text-decoration:underline;}
 
 /**** CONTENT *******/
-article							{width:800px; cursor:default;}
+article							{width:850px; cursor:default;}
```
Pas favorable à élargir la largeur par défaut, à cause du support de la largeur la plus répandue sur les supports mobiles.

FlyinPhil replied:
Ok, je vais essayer de trouver une autre manière d'afficher les deux valeurs sans élargir. J'utilise l'appli à 850 de largeur sur mobile sans souci , car de toute façon je suis obligé de zoomer pour voir les infos .

toxcct replied:
Je crois qu'il va falloir que je repense l'UI pour les mobiles de toutes façons, mais je ne sais pas trop comment m'y prendre pour ce nouveau design :/

toxcct commented on js/polars_app.js:
```diff
@@ -48,6 +48,7 @@ var PolarsApp = (function(){
 		}
 	};
 	var _acceptedURLParams	= ["race_id", "twa", "tws", "light", "reach", "heavy", "foil", "hull", "utm_source"];
+	const sailTolerance = 1.014 ;  //  "autoSailChangeTolerance" in polar json, the same for all boats.
```
Serait-il possible d'extraire cette valeur des polaires, même si elle est invariante depuis sa création?

FlyinPhil replied:
Vu la manière dont sont récupérées les valeurs des polaires, sauf erreur bien compliqué pour une valeur constante depuis deux ans et pour tous les bateaux.

toxcct replied:
Les données du JSON des polaires sont lues intégralement dans la méthode `PolarsReader.getSpeeds()` ([polars_reader.js:33](https://github.com/toxcct/VRPolarsChart/blob/main/js/polars_reader.js#L33)).
Un champ supplémentaire pourrait être retourné par cette méthode pour renvoyer la valeur de cette tolérance.

Au passage, je note qu'il existe plusieurs _tolérances_ différentes dans les polaires (`autoSailChangeTolerance` et `badSailTolerance`) ; connait-on l'usage de cette autre valeur de tolérance qui semble concerner les mauvaises voiles ?

FlyinPhil replied:
Le "badSailTolerance" était auparavant la tolérance qui faisait changer de voile quand on était en voile automatique, maintenant ce coefficient permet de définir le seul d'alerte de l'utilisateur qui n'a pas la voile la plus favorable et qui est en voile manuelle.

toxcct commented on js/polars_app.js:
```diff
@@ -1090,13 +1091,13 @@ var PolarsApp = (function(){
-			$(".sail.Spi"		).next().next().text(data.sailsSpeeds.Spi		.fix(2, true));
-			$(".sail.LightGnk"	).next().next().text(data.sailsSpeeds.LightGnk	.fix(2, true));
-			$(".sail.HeavyGnk"	).next().next().text(data.sailsSpeeds.HeavyGnk	.fix(2, true));
+			$(".sail.Jib"		).next().next().text(data.sailsSpeeds.Jib		.fix(3, true) + ( (					  (data.sailsSpeeds.Jib      * sailTolerance > _current_boatspeed) && (data.sailsSpeeds.Jib      !=_current_boatspeed)) ? (' (' +_current_boatspeed.fix(3, true) + ')' ) : "")  )  ;
+			$(".sail.LightJib"	).next().next().text(data.sailsSpeeds.LightJib	.fix(3, true) + ( ( _chk_opt_light && (data.sailsSpeeds.LightJib * sailTolerance > _current_boatspeed) && (data.sailsSpeeds.LightJib !=_current_boatspeed)) ? (' (' +_current_boatspeed.fix(3, true) + ')' ) : "")  )  ;
+			$(".sail.Staysail"	).next().next().text(data.sailsSpeeds.Staysail	.fix(3, true) + ( ( _chk_opt_heavy && (data.sailsSpeeds.Staysail * sailTolerance > _current_boatspeed) && (data.sailsSpeeds.Staysail !=_current_boatspeed)) ? (' (' +_current_boatspeed.fix(3, true) + ')' ) : "")  )  ;
+			$(".sail.Code0"		).next().next().text(data.sailsSpeeds.Code0		.fix(3, true) + ( ( _chk_opt_reach && (data.sailsSpeeds.Code0	 * sailTolerance > _current_boatspeed) && (data.sailsSpeeds.Code0    !=_current_boatspeed)) ? (' (' +_current_boatspeed.fix(3, true) + ')' ) : "")  )  ;
+			$(".sail.Spi"		).next().next().text(data.sailsSpeeds.Spi		.fix(3, true) + ( (					  (data.sailsSpeeds.Spi	 	 * sailTolerance > _current_boatspeed) && (data.sailsSpeeds.Spi	     !=_current_boatspeed)) ? (' (' +_current_boatspeed.fix(3, true) + ')' ) : "")  )  ;
+			$(".sail.LightGnk"	).next().next().text(data.sailsSpeeds.LightGnk	.fix(3, true) + ( ( _chk_opt_light && (data.sailsSpeeds.LightGnk * sailTolerance > _current_boatspeed) && (data.sailsSpeeds.LightGnk !=_current_boatspeed)) ? (' (' +_current_boatspeed.fix(3, true) + ')' ) : "")  )  ;
+			$(".sail.HeavyGnk"	).next().next().text(data.sailsSpeeds.HeavyGnk	.fix(3, true) + ( ( _chk_opt_heavy && (data.sailsSpeeds.HeavyGnk * sailTolerance > _current_boatspeed) && (data.sailsSpeeds.HeavyGnk !=_current_boatspeed)) ? (' (' +_current_boatspeed.fix(3, true) + ')' ) : "")  )  ;
```
Pas favorable à passer l'affichage à 3 digits à cet endroit.
Il ne s'agit que d'affichage et les calculs prennent bien en compte la précision maximale disponible. A la rigueur, on peut rajouter une infobulle qui affiche une valeur plus précise quand la souris passe dessus...

---

Concernant le changement fonctionnel, serait-il possible de me réexpliquer le comportement de VR par rapport à ce coefficient de tolérance ?
Est ce que la vitesse est _augmentée_ sans pour autant changer la voile active ?
Le cas échéant, pourquoi ne pas simplement remplacer la vitesse théorique (sans ce coeff) par la vitesse augmentée ?

FlyinPhil replied:
La vitesse du bateau est augmentée et utilise la vitesse de la voile la plus favorable si on est en voile auto et dans la tolérance de 1.014 (1.4%) ). Si on dépasse cette tolérance on change de voile. Pas mal de joueurs demandent à retirer cette fonctionnalité,  incohérente avec le monde réel et à remettre , comme avant, un seuil de (0.3%) avant de changer de voile automatiquement. Ok pour ne mettre que 2 digits + infobulle. j'en ai mis un 3ème car parfois on se demande pourquoi une voile est en gras alors que les valeurs semblent égales avec deux digits.

toxcct replied:
Ok, donc le bateau ne gagne pas l'équivalent de la totalité de la tolérance en bonus dans le cas de recouvrement de voiles, mais il applique la vitesse de la meilleure voile. Donc à ce niveau-là, c'est déjà ce que faisait mon appli, puisqu'elle trace la courbe des meilleures vitesses.
Du coup, peux-tu m'expliquer ce qu'apporte cette nouvelle valeur affichée entre parenthèses ?

Question : Est-ce que ce bonus s'applique également quand on n'est pas en voiles auto, sans qu'il puisse changer de voile une fois que la tolérance est dépassée ?

FlyinPhil replied:
Ok, l'appli traçait bien la courbe des meilleures vitesses mais ne permettait pas de savoir quelles voiles avaient une vitesse dans le seuil de tolérance et profitaient d'une meilleure vitesse. C'est juste l'info qu'il manque, car dans le jeu, ce n'est pas uniquement le gain en vitesse qui nous intéresse mais aussi et surtout de savoir si on va changer de voile (et se prendre les pénalités correspondantes). 
Plutôt que de mettre la meilleure vitesse entre parenthèse et pour gagner en largeur, on peut choisir de juste mettre un "(B)"  (pour Boost) ou "(T)" (pour Tolérance) derrière la vitesse de la voile et en info bulle le coefficient par rapport à la meilleure voile et/ou la meilleure vitesse.

toxcct commented on css/general.css:
```diff
@@ -75,7 +75,7 @@ article .sidebox.grid5cols	> .title	{grid-column:span 5;}
 article .sidebox.grid2cols	> .sep		{grid-column:span 2;}
 article .sidebox.grid5cols	> .sep		{grid-column:span 5;}
-article .sidebox.grid5cols	> input		{grid-column:span 3; text-align:center; vertical-align:middle; height:18px; border:1px solid #999; padding:0;}
+article .sidebox.grid5cols	> input		{grid-column:span 2; text-align:center; vertical-align:middle; height:18px; border:1px solid #999; padding:0;}
```
```suggestion
article .sidebox.grid5cols	> input		{grid-column:span 3; text-align:center; vertical-align:middle; height:18px; border:1px solid #999; padding:0; width:calc(100% - 2px);}
```

FlyinPhil committed:
Update index.htm

Co-authored-by: toxcct <toxcct@users.noreply.github.com>

FlyinPhil committed:
Update css/general.css

Co-authored-by: toxcct <toxcct@users.noreply.github.com>

FlyinPhil committed:
Update css/general.css

Co-authored-by: toxcct <toxcct@users.noreply.github.com>

FlyinPhil committed:
_autoSailChangeTolerance

Récupération de ce paramètre dans les polaires. Et changement de l'affichage de l'effet boost dans la colonne

FlyinPhil reviewed (commented):
Le texte que je met en "title" : "En voile automatique, boost coef x.xxxx (peut profiter de la vitesse de la meilleure voile)" mérite sûrement d'être réduit et revu.  "En voile automatique, boost coef x.xxxx (profite de la vitesse de la meilleure voile)".. ou encore plus court
